refactor: WIP-1001 abstract account model with pluggable admin authorities (#587)
worldcoin/world-chain: wips/wip-1001.md

---
wip: 1001
title: World Chain Native Account Abstraction

> description: A precompile-managed account model with a single admin authority pinned at creation, supporting WorldID, Secp256k1, and P256 admins. Session keys sign EIP-2718 typed transactions (`0x1D`) on the account; admin authorities gate session-key set mutations.

Reviewer: ```suggestion
description: A smart contract managed account model with a single admin authority pinned at creation, supporting WorldID, Secp256k1, and P256 admins. Session keys sign EIP-2718 typed transactions (`0x1D`) on the account; admin authorities gate session-key set mutations.
```

## Abstract

> A *World Chain Account* is a precompile-managed account whose 20-byte address is deterministic and lifetime-stable. Each account has a single *admin authority* fixed at creation, which gates mutation of the account's *Keyring* — the set of session keys authorized to sign on its behalf, held natively in the precompile and read by the protocol during transaction validation. We extend [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) with a typed transaction envelope (`0x1D`) signed by any session key in a *World Chain Account*'s keyring.

Reviewer: ```suggestion
A *World Chain Account* is a precompile-managed account whose 20-byte address is deterministic and lifetime-stable. Each account has a single *admin authority* fixed at creation, which gates mutation of the account's *Keyring* — the set of session keys authorized to sign on its behalf. We extend [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) with a typed transaction envelope (`0x1D`) signed by any session key in a *World Chain Account*'s keyring.
```

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).

### Constants

| Name                              | Value                                        | Description                                                 |
| --------------------------------- | -------------------------------------------- | ----------------------------------------------------------- |
| `WORLD_TX_TYPE`                   | `0x1D`                                       | [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) tx type |
| `WORLD_CHAIN_ACCOUNT_PRECOMPILE`  | `0x000000000000000000000000000000000000001D` | Stateful precompile managing the World Chain Account registry |
| `MAX_SESSION_KEYS`                | `20`                                         | Maximum authorized session keys per account                 |

> | `ADMIN_TYPE_WORLD_ID`             | `0x00`                                       | Admin type discriminator: WorldID admin                     |

Reviewer: Prefix these constants with `WORLD_CHAIN_ACCOUNT`

Reviewer: Also what happened to `WORLD_CHAIN_RP_ID`

Author: WORLD_CHAIN_RP_ID was removed because it is a specific to the world id admin key type and does not belong here

| `WebAuthn`  | `0x02` | [secp256r1](https://neuromancer.sk/std/nist/P-256) (NIST P-256)                         | 64 bytes — uncompressed `(x, y)` (P256 under [WebAuthn](https://www.w3.org/TR/webauthn-3/)) |
| `EdDSA`     | `0x03` | [edwards25519](https://datatracker.ietf.org/doc/html/rfc8032#section-5.1)               | 32 bytes — compressed public key (Ed25519)                                                  |

> `KeyType` and `adminType` occupy disjoint domains: their numeric values are defined in different enums and MUST NOT be used interchangeably in the precompile interface or the address-derivation preimage. Session keys carry no intrinsic privileges beyond signing `0x1D` transactions; per-key permissions (expiry, spend limits, call scopes) are out of scope for this WIP (see [Extensions](#extensions)).

Reviewer: Sure, but I don't think we need to make the terms distinct here since the sets of key types are identical

}
```

- **`Create`** — installs `keys` as the full keyring at the account's address and initializes the account with the supplied admin authority. MUST revert if the account already exists, if `keys.length == 0`, or if `keys.length > MAX_SESSION_KEYS`.
- **`Add`** — appends `keys` to the existing keyring. MUST revert if the account does not exist, if any key is already present, or if the resulting keyring would exceed `MAX_SESSION_KEYS`.
- **`Remove`** — removes each entry in `keys` from the existing keyring. MUST revert if the account does not exist or if any key is absent. The keyring MAY become empty; an account with zero session keys exists but cannot sign transactions until the next `Add`.

The `signalHash` an admin authority binds to is uniform across instantiations.

```solidity
struct CreateSignal {

>     bytes32       tag;             // == keccak256("WIP-1001/create")

Reviewer: ```suggestion
    bytes32       tag;             // == keccak256("WORLD_CHAIN_ACCOUNT_CREATE")
```

>     address       msgSender;

Reviewer: Why is this necessary?

Reviewer: Also this is replayable?

    KeyringUpdate keyringUpdate;
}

struct UpdateSignal {

>     bytes32       tag;             // == keccak256("WIP-1001/update")

Reviewer: ```suggestion
    bytes32       tag;             // == keccak256("WORLD_CHAIN_ACCOUNT_UPDATE")
```

    address       accountAddress;

>     uint64        adminNonce;

Reviewer: Ah
